File: lib/services/currency_recognition_service.dart
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:tflite_flutter/tflite_flutter.dart';
import 'package:image/image.dart' as img;
import 'package:ningapi/models/recognition_result.dart';

class CurrencyRecognitionService {
  static final CurrencyRecognitionService _instance = CurrencyRecognitionService._internal();
  factory CurrencyRecognitionService() => _instance;
  CurrencyRecognitionService._internal();

  Interpreter? _interpreter;
  List<String> _labels = [];
  bool _isInitialized = false;

  // Taille d'entree du modele (MobileNetV2 utilise 224x224)
  int _inputSize = 224;
  static const int numChannels = 3;
  
  /// Initialise le modele et charge les labels
  Future<void> initialize() async {
    if (_isInitialized) return;

    try {
      _interpreter = await Interpreter.fromAsset('assets/models/model.tflite');
      debugPrint('Modele TFLite charge');

      // Detection dynamique de la taille d'entree
      final inputTensor = _interpreter!.getInputTensor(0);
      _inputSize = inputTensor.shape[1];
      debugPrint('Taille entree detectee: $_inputSize x $_inputSize');

      // Labels specifiques au modele (ordre alphabetique comme dans l'entrainement)
      _labels = [
        '1\$', '10\$', '100\$',
        '10000FC', '1000FC', '100FC',
        '20\$', '20000FC', '200FC',
        '5\$', '50\$',
        '5000FC', '500FC', '50FC'
      ];
      debugPrint('Labels charges: $_labels');

      _isInitialized = true;
    } catch (e) {
      debugPrint('Erreur initialisation: $e');
      rethrow;
    }
  }

  /// Effectue la reconnaissance de billet sur une image
  Future<CurrencyResult> recognizeCurrency(File imageFile) async {
    if (!_isInitialized) {
      await initialize();
    }

    try {
      final processedImage = await _preprocessImage(imageFile);
      final input = [processedImage];
      final output = List.filled(1, List.filled(_labels.length, 0.0)).cast<List<double>>();

      _interpreter!.run(input, output);

      final probabilities = output[0];
      
      int maxIndex = 0;
      double maxProb = probabilities[0];
      debugPrint('Probabilites: $probabilities');

      for (int i = 1; i < probabilities.length; i++) {
        if (probabilities[i] > maxProb) {
          maxProb = probabilities[i];
          maxIndex = i;
        }
      }
      
      debugPrint('Resultat: ${_labels[maxIndex]} ($maxProb)');

      final allProbs = <String, double>{};
      for (int i = 0; i < _labels.length; i++) {
        allProbs[_labels[i]] = probabilities[i];
      }

      final label = _labels[maxIndex];
      String currency = 'FC';
      if (label.toUpperCase().contains('USD') || label.contains('\$')) {
        currency = 'USD';
      }

      return CurrencyResult(
        denomination: label,
        currency: currency,
        confidence: maxProb,
        timestamp: DateTime.now(),
        allProbabilities: allProbs,
      );
    } catch (e) {
      debugPrint('Erreur reconnaissance: $e');
      rethrow;
    }
  }

  Future<List<List<List<double>>>> _preprocessImage(File imageFile) async {
    final bytes = await imageFile.readAsBytes();
    
    img.Image? image = img.decodeImage(bytes);
    if (image == null) {
      throw Exception('Impossible de decoder image');
    }

    // Corriger l'orientation (EXIF)
    image = img.bakeOrientation(image);

    img.Image resized = img.copyResize(
      image,
      width: _inputSize,
      height: _inputSize,
      interpolation: img.Interpolation.linear,
    );

    // Pretraitement MobileNetV2: normalisation dans [-1, 1]
    // Equivalent de tf.keras.applications.mobilenet_v2.preprocess_input
    // Formule: (pixel / 127.5) - 1.0
    List<List<List<double>>> processedImage = List.generate(
      _inputSize,
      (y) => List.generate(
        _inputSize,
        (x) {
          final pixel = resized.getPixel(x, y);
          return [
            (pixel.r / 127.5) - 1.0,
            (pixel.g / 127.5) - 1.0,
            (pixel.b / 127.5) - 1.0,
          ];
        },
      ),
    );

    return processedImage;
  }

  void dispose() {
    _interpreter?.close();
    _isInitialized = false;
  }
}

File: .OTLIN/services/CurrencyRecognitionService.kt
package com.ningapi.services

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import com.ningapi.models.CurrencyResult
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.support.common.FileUtil
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.time.LocalDateTime

class CurrencyRecognitionService private constructor(private val context: Context) {

    companion object {
        @Volatile
        private var INSTANCE: CurrencyRecognitionService? = null

        fun getInstance(context: Context): CurrencyRecognitionService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: CurrencyRecognitionService(context).also { INSTANCE = it }
            }
        }
        
        private const val TAG = "CurrencyRecognition"
    }

    private var interpreter: Interpreter? = null
    // Labels specifiques au modele (ordre alphabetique comme dans l'entrainement)
    private val labels = listOf(
        "1$", "10$", "100$",
        "10000FC", "1000FC", "100FC",
        "20$", "20000FC", "200FC",
        "5$", "50$",
        "5000FC", "500FC", "50FC"
    )
    
    private var isInitialized = false
    private var inputSize = 224

    suspend fun initialize() {
        if (isInitialized) return

        try {
            // Load model
            // Assuming model.tflite is in assets/models/
            val mappedByteBuffer = FileUtil.loadMappedFile(context, "models/model.tflite")
            interpreter = Interpreter(mappedByteBuffer)
            Log.d(TAG, "Modele TFLite charge")

            // Detection dynamique de la taille d'entree
            val inputTensor = interpreter!!.getInputTensor(0)
            inputSize = inputTensor.shape()[1]
            Log.d(TAG, "Taille entree detectee: $inputSize x $inputSize")
            Log.d(TAG, "Labels charges: $labels")

            isInitialized = true
        } catch (e: Exception) {
            Log.e(TAG, "Erreur initialisation: $e", e)
            throw e
        }
    }

    suspend fun recognizeCurrency(imageFile: File): CurrencyResult {
        if (!isInitialized) {
            initialize()
        }

        try {
            val bitmap = BitmapFactory.decodeFile(imageFile.absolutePath)
                ?: throw Exception("Impossible de decoder image")

            // Preprocess
            val inputBuffer = preprocessImage(bitmap)
            
            // Output buffer: [1, labels.size]
            val outputBuffer = Array(1) { FloatArray(labels.size) }
            
            interpreter!!.run(inputBuffer, outputBuffer)

            val probabilities = outputBuffer[0]
            
            var maxIndex = 0
            var maxProb = probabilities[0]
            
            Log.d(TAG, "Probabilites: ${probabilities.joinToString()}")

            for (i in 1 until probabilities.size) {
                if (probabilities[i] > maxProb) {
                    maxProb = probabilities[i]
                    maxIndex = i
                }
            }

            Log.d(TAG, "Resultat: ${labels[maxIndex]} ($maxProb)")

            val allProbs = mutableMapOf<String, Double>()
            for (i in labels.indices) {
                allProbs[labels[i]] = probabilities[i].toDouble()
            }

            val label = labels[maxIndex]
            var currency = "FC"
            if (label.uppercase().contains("USD") || label.contains("$")) {
                currency = "USD"
            }

            return CurrencyResult(
                denomination = label,
                currency = currency,
                confidence = maxProb.toDouble(),
                timestamp = LocalDateTime.now(),
                allProbabilities = allProbs
            )

        } catch (e: Exception) {
            Log.e(TAG, "Erreur reconnaissance: $e", e)
            throw e
        }
    }

    private fun preprocessImage(bitmap: Bitmap): ByteBuffer {
        // Resize
        val resized = Bitmap.createScaledBitmap(bitmap, inputSize, inputSize, true)

        // Allocate ByteBuffer
        // 4 bytes per float * 3 channels * width * height
        val byteBuffer = ByteBuffer.allocateDirect(4 * 3 * inputSize * inputSize)
        byteBuffer.order(ByteOrder.nativeOrder())

        val intValues = IntArray(inputSize * inputSize)
        resized.getPixels(intValues, 0, resized.width, 0, 0, resized.width, resized.height)

        // Normalize [-1, 1]
        // (pixel / 127.5) - 1.0
        for (pixelValue in intValues) {
            // Extract RGB
            val r = (pixelValue shr 16) and 0xFF
            val g = (pixelValue shr 8) and 0xFF
            val b = pixelValue and 0xFF

            // MobileNetV2 preprocessing
            byteBuffer.putFloat((r / 127.5f) - 1.0f)
            byteBuffer.putFloat((g / 127.5f) - 1.0f)
            byteBuffer.putFloat((b / 127.5f) - 1.0f)
        }
        
        return byteBuffer
    }

    fun dispose() {
        interpreter?.close()
        isInitialized = false
    }
}
